pkg/cmd/pulumi/preview.go

```diff
@@ -18,22 +18,152 @@ import (
 	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
 )
 
+// buildImportFile takes an event stream from the engine and builds an import file from it for every create.
+func buildImportFile(events <-chan engine.Event) *promise.Promise[importFile] {
+	return promise.Run(func() (importFile, error) {
+		// A mapping of every URN we see to it's name, used to build later resourceSpecs
+		fullNameTable := map[resource.URN]string{}
+		// A set of all URNs we've added to the import list, used to avoid adding parents to NameTable.
+		importSet := map[resource.URN]struct{}{}
+		// All providers that we've seen so far, used to build Version and PluginDownloadURL.
+		providerInputs := map[resource.URN]resource.PropertyMap{}
+
+		imports := importFile{
+			NameTable: map[string]resource.URN{},
+		}
+
+		// This is a pretty trivial mapping of Create operations to import declarations.
+		for e := range events {
+			preEvent, ok := e.Payload().(engine.ResourcePreEventPayload)
+			if !ok {
+				continue
+			}
+
+			urn := preEvent.Metadata.URN
+			fullNameTable[urn] = urn.Name()
+
+			// If this is a provider we need to note we've seen it so we can build the Version and PluginDownloadURL of
+			// any resources that use it.
+			if providers.IsProviderType(urn.Type()) {
+				providerInputs[urn] = preEvent.Metadata.Res.Inputs
+			}
+
+			// Only interested in creates
+			if preEvent.Metadata.Op != deploy.OpCreate {
+				continue
+			}
+			// No need to import the root stack even if it needs creating
+			if preEvent.Metadata.Type == resource.RootStackType {
+				continue
+			}
+
+			// We're importing this URN so track that we've seen it.
+			importSet[urn] = struct{}{}
+
+			// We can't actually import providers yet, just skip them. We'll only error if anything
+			// actually tries to use it.
+			if providers.IsProviderType(urn.Type()) {
+				continue
+			}
+
+			new := preEvent.Metadata.New
+			contract.Assertf(new != nil, "%s: expected new resource for a create to be non-nil", urn)
+
+			var parent string
+			if new.Parent != "" {
+				// If the parent is just the root stack then skip it as we don't need to import that.
+				if new.Parent.Type() != resource.RootStackType {
+					var has bool
+					parent, has = fullNameTable[new.Parent]
+					contract.Assertf(has, "expected parent %q to be in full name table", new.Parent)
+					// Don't add to the import NameTable if we're importing this in the same deployment.
+					if _, has := importSet[new.Parent]; !has {
+						imports.NameTable[parent] = new.Parent
+					}
+				}
+			}
+
+			var provider, version, pluginDownloadURL string
+			if new.Provider != "" {
+				ref, err := providers.ParseReference(new.Provider)
+				if err != nil {
+					return importFile{}, fmt.Errorf("could not parse provider reference: %v", err)
+				}
+
+				// If we're trying to create this provider in the same deployment and it's not a default provider then
+				// we need to error, the import system can't yet "import" providers.
```

Comment:
Is there an issue tracking this?

Reply:
What does this mean in practice for users that have explicit providers in their stack? The resources that use that explicit provider will be importable? But the explicit providers themselves aren't importable?

Reply:
It errors that the provider needs to be created first then they can import it. See a few lines down "cannot import resource %q with a new explicit provider %q".